microcloud/charms/src/charm.py

```diff
@@ -0,0 +1,677 @@
+        """Initialize charm's variable."""
+        super().__init__(*args)
+
+        # Initialize the persistent storage if needed
+        self._stored.set_default(
+            config={},
+            microcloud_binary_path="",
+            microcloud_snap_path="",
+        )
+
+        # Main event handlers
+        self.framework.observe(self.on.install, self._on_charm_install)
+        self.framework.observe(self.on.config_changed, self._on_charm_config_changed)
+        self.framework.observe(self.on.start, self._on_charm_start)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+        # Relation event handlers
+        self.framework.observe(self.on.cluster_relation_created, self._on_cluster_relation_created)
+        self.framework.observe(self.on.cluster_relation_joined, self._on_cluster_relation_joined)
+        self.framework.observe(
+            self.on.cluster_relation_departed, self._on_cluster_relation_departed
+        )
+        self.framework.observe(self.on.cluster_relation_broken, self._on_cluster_relation_broken)
+
+    @property
+    def peers(self):
+        """Fetch the cluster relation."""
+        return self.model.get_relation("cluster")
+
+    def get_peer_data_str(self, bag, key: str) -> str:
+        """Retrieve a str from the peer data bag."""
+        if not self.peers or not bag or not key:
+            return ""
+
+        value = self.peers.data[bag].get(key, "")
+        if isinstance(value, str):
+            return value
+
+        logger.error(f"Invalid data pulled out from {bag.name}.get('{key}')")
+        return ""
+
+    def set_peer_data_str(self, bag, key: str, value: str) -> None:
+        """Put a str into the peer data bag if not there or different."""
+        if not self.peers or not bag or not key:
+            return
+
+        old_value: str = self.get_peer_data_str(bag, key)
+        if old_value != value:
+            self.peers.data[bag][key] = value
+
+    def _on_charm_install(self, event: InstallEvent) -> None:
+        logger.info("Installing the Microcloud charm")
+        # Confirm that the config is valid
+        if not self.config_is_valid():
+            return
+
+        # Install Microcloud itself
+        try:
+            self.snap_install_microcloud()
+            logger.info("Microcloud installed successfully")
+        except RuntimeError:
+            logger.error("Failed to install Microcloud")
+            event.defer()
+            return
+
+        # Apply sideloaded resources attached at deploy time
+        self.resource_sideload()
+
+    def _on_cluster_relation_created(self, event: RelationCreatedEvent) -> None:
+        """We must wait for all units to be ready before initializing Microcloud."""
+        self.set_peer_data_str(self.unit, "new_node", "True")
+        if not self.unit.is_leader():
+            time.sleep(30)  # Make sure that the leader has time to initialize the cluster
+
+        return
```

Comment:
I'm not sure delaying the return achieves anything, or does it? Isn't the leader supposed to bootstrap the cluster and then only add nodes?